AppKit/CPWindow/CPWindow.j

```diff
@@ -3844,6 +3819,393 @@ var interpolate = function(fromValue, toValue, progress)
 
 @end
 
+@implementation CPWindow (TrackingAreaAdditions)
+
+- (void)_addTrackingAreaView:(CPView)aView
+{
+    var trackingAreas = [aView trackingAreas];
+    
+    for (var i = 0; i < trackingAreas.length; i++)
+        [self _addTrackingArea:trackingAreas[i]];
+}
+
+- (void)_removeTrackingAreaView:(CPView)aView
+{
+    var trackingAreas = [aView trackingAreas];
+    
+    for (var i = 0; i < trackingAreas.length; i++)
+        [self _removeTrackingArea:trackingAreas[i]];
+}
+
+- (void)_addTrackingArea:(CPTrackingArea)trackingArea
+{
+    var trackingAreaView = [trackingArea view];
+    
+    if (![_trackingAreaViews containsObject:trackingAreaView])
+        [_trackingAreaViews addObject:trackingAreaView];
+    
+    // If CPTrackingAssumeInside option is set, put the tracking area in the _mouseEnteredStack
+    
+    if ([trackingArea options] & CPTrackingAssumeInside)
+        [_mouseEnteredStack addObject:trackingArea];
+}
+
+- (void)_removeTrackingArea:(CPTrackingArea)trackingArea
+{
+    // If mouse is in the tracking area, we remove it from the stack to avoid to fire a future mouseExited event
+    
+    [_mouseEnteredStack removeObjectIdenticalTo:trackingArea];
+    
+    var trackingAreaView = [trackingArea view];
+    
+    [_trackingAreaViews removeObjectIdenticalTo:trackingAreaView];
+}
+
+- (void)_handleTrackingAreaEvent:(CPEvent)anEvent
```

Comment:
I agree at 100% to your concerns about readability and maintainability (that's why my code is generally a little verbose) and to your remarks.

If you combine this with performance objective, shouldn't we apply the following policy : each layer of code should use lower layers respecting APIs (this is for readability), and should optimize its own implementation.
This policy will not give the optimum performances but represents a good balance between both objectives. (this is what I ask my own developers to do)
In Cappuccino, this would result, for example, in implementing your suggestion (anArray.indexOf(object) >= 0) for "containsObject" method. And to use [anArray containsObject:anObject] everywhere above CPArray. 
Or, another example, to use for loops the traditional "for (var i = 0, count = [anArray count]; i < count; i++)" instead of "for (var i = 0; i < anArray.length; i++)".

Do you agree with this policy ?

If so, shouldn't we propose some update to base classes (if possible, of course) ?

And in the case in point, I'll refactor the code this weekend.